JetBrains: fix search combobox styles

taras-yemets opened the pull request:
Closes https://github.com/sourcegraph/sourcegraph/issues/37689

## Test plan
Tested visually. 

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

vdavid commented on client/search-ui/src/input/SearchContextMenu.tsx:
```diff
@@ -311,6 +312,7 @@ export const SearchContextMenu: React.FunctionComponent<React.PropsWithChildren<
                     onKeyDown={onCloseButtonKeyDown}
                     variant="icon"
                     className={styles.titleClose}
+                    data-close-button={true}
```
I see that this and two more changes are outside of the JetBrains codebase. Have you made sure that these changes lead to no regressions?

taras-yemets replied:
Thank you for your review, @vdavid!
I double-checked and didn't spot any regressions. But anyway I refactored this logic to not need this data attribute for JetBrains plugin styling.

vdavid replied:
Thanks! Feel free to merge it then :)

vdavid reviewed (approved):
The Loom is very helpful and the new style looks great!
The code looks good to me **if** the answer to my question (did you ensure the changes cause no regressions in our other products?) is “yes”.
Getting an approval from Sara on the Loom might be an extra layer of reassurance. @jjinnii if you have a minute, please take a look at the [Loom](https://www.loom.com/share/f8110a2a28754282a15a2264c481390d).

philipp-spiess commented:
@taras-yemets Amazing work! I only looked at the screenshots to be fair 🙈 

I spotted one thing that I think we can improve:

![image](https://user-images.githubusercontent.com/458591/176430190-7caf27ff-6981-4cb0-88aa-791feb6506dc.png)

In this screenshot, the contrast between the black text (`ohascomittafter`) and the blue background color makes it hard to read the text. Do you think we can customize it to be a bit brighter?

taras-yemets commented:
@philipp-spiess, thank you for your review!

 >Do you think we can customize it to be a bit brighter?
 
 Improved colors contrasts in https://github.com/sourcegraph/sourcegraph/pull/37833/commits/c87c597bf3f34f7daab6191fba7b49566c186fae
 

<img width="971" alt="Screenshot 2022-06-29 at 15 38 06" src="https://user-images.githubusercontent.com/25318659/176438323-d6963305-98cf-489d-911e-a456ae7fc831.png">
<img width="971" alt="Screenshot 2022-06-29 at 15 37 43" src="https://user-images.githubusercontent.com/25318659/176438335-8a8bf660-9942-4b6c-bddd-7c44f1ec76e6.png">

philipp-spiess commented:
@taras-yemets I don't see a big difference in the screenshots 🙁  This part is still barely readable:

![Screenshot 2022-06-29 at 14 54 55](https://user-images.githubusercontent.com/458591/176441414-84202c82-0835-4048-9001-bd9b581405d5.png)

What's the calculated contrast that Chrome shows when you use the inspector tool?

![Screenshot 2022-06-29 at 14 56 31](https://user-images.githubusercontent.com/458591/176441598-0247d13d-935e-4e11-bc0f-7a014d827710.png)

taras-yemets commented:
@philipp-spiess, my bad! I updated the contrast of the detail text https://github.com/sourcegraph/sourcegraph/commit/c87c597bf3f34f7daab6191fba7b49566c186fae, but not of the main text you pointed to.

<img width="971" alt="Screenshot 2022-06-29 at 16 02 21" src="https://user-images.githubusercontent.com/25318659/176442931-8b203cb0-06a7-4496-b86e-7f76dc4f73ab.png">

On it. Sorry for the confusion!

philipp-spiess commented:
> @taras-yemets I don't see a big difference in the screenshots 🙁 This part is still barely readable:

Oh I haven't even noticed this part, thanks for fixing that! 😊

philipp-spiess commented:
> What do you think, are we OK with not highlighting the matched text of the selected item?

Yeah I think this makes sense but I’m also curious what @jjinnii thinks 🙂  Thanks so much for looking into it!

jjinnii commented:
Thanks everyone for the teamwork on this! I have a couple questions and thoughts: 

- **Question**: In the first Before/After screenshots, it seems as if this issue also tackles getting rid of the blue border when the searchbox is in active state, related to #37285, but I don't see it consistent through the Loom. What changes were the first before/after screenshots meant to reflect? 
- **Question**: Is the activate row color change the only change made to the context menu? And is there a reason we've introduced a grey border around the searchfield inside the context menu?
<img width="842" alt="Screen Shot 2022-06-30 at 5 13 32 PM" src="https://user-images.githubusercontent.com/87138876/176627599-b8e0137d-1650-4b87-966b-e56e87d78779.png">

- **Thought**: On the discussion about the highlighted colors in the auto-complete dropdown (Thanks @philipp-spiess for calling this out!) My suggestion would be to solve this the way JetBrains solves it in their UI (screenshot below). This is basically the solution that you all explored (white text), but we maintain the highlighting of the searched value with their native highlighted yellow. @taras-yemets I wasn't sure thought by "By making it white we no longer highlight the matched text of the selected item", if you meant this solution is not technically feasible. Let me know if this is the case.
<img width="701" alt="Screen Shot 2022-06-30 at 5 15 51 PM" src="https://user-images.githubusercontent.com/87138876/176628005-4ef31893-d72c-4fa1-af09-6cfe10f2d780.png">

- **Though**t: Apologies as I can't inspect the CSS on this. The font being used in the auto-complete dropdown seems to be different from JetBrains native fonts, can we check this? The font weight also seems to be much heavier than what I see being used in "Find in Files too"
- **Question**: A question for the engineers who use IntelliJ. What do we think about not using the Sourcegraph style "Tab button and just having the helper text on the keyboard shortcut directly to the right in bold?
<img width="693" alt="Screen Shot 2022-06-30 at 6 21 18 PM" src="https://user-images.githubusercontent.com/87138876/176641651-4a7c12e2-965c-4e97-b51d-c3dbac508d2b.png">

taras-yemets commented:
Thank you for your review, @jjinnii!

 >Question: In the first Before/After screenshots, it seems as if this issue also tackles getting rid of the blue border when the searchbox is in active state, related to https://github.com/sourcegraph/sourcegraph/issues/37285, but I don't see it consistent through the Loom. What changes were the first before/after screenshots meant to reflect?
 
Sorry for the confusion, the issue with the focused item outline is not addressed in the current PR. The first before/after screenshots reflect the tiny changes in the context menu button. Although, I changed it a bit later by changing the background, hover and focus styles, and outline. 

>Question: Is the activate row color change the only change made to the context menu? And is there a reason we've introduced a grey border around the searchfield inside the context menu?

Changes to the context menu: removed border-radius on container and input, changed active row bg colo, adjusted input bg color, and outline. The grey borders were preserved by mistake from the original component (I just changed color), now removed. 

<details>
<summary><b>Screenshots</b></summary>
<table>
<tr>
<td><img width="1024" alt="Screenshot 2022-06-30 at 23 10 21" src="https://user-images.githubusercontent.com/25318659/176768933-00b3527b-c67d-4064-9a36-dfc3229cec85.png"></td>
<td><img width="1024" alt="Screenshot 2022-06-30 at 23 10 07" src="https://user-images.githubusercontent.com/25318659/176768950-b7a65a9c-b13c-4e47-87b3-8b8001cc2d06.png"></td>
</tr>
</table>
</details>

> Thought: On the discussion about the highlighted colors in the auto-complete dropdown ... maintain the highlighting of the searched value with their native highlighted yellow

I somehow missed this option, thank you! Implemented the match query highlight in yellow. 

<details>
<summary><b>Screenshots</b></summary>
<table>
<tr>
<td>
<img width="1024" alt="Screenshot 2022-06-30 at 23 14 04" src="https://user-images.githubusercontent.com/25318659/176770322-a15231c6-6c97-426f-9c18-3c089fb01bad.png">
</td>
<td>
<img width="1024" alt="Screenshot 2022-06-30 at 23 14 27" src="https://user-images.githubusercontent.com/25318659/176770373-c46478b2-d497-4f83-958c-5a91f905763e.png">
</td>
</tr>
</table>
</details>

> Thought: Apologies as I can't inspect the CSS on this. The font being used in the auto-complete dropdown seems to be different from JetBrains native fonts, can we check this? The font weight also seems to be much heavier than what I see being used in "Find in Files too"

Nice catch! I updated the dropdown menu font family and font-weight to be the same as in the JetBrains UI (see screenshots in the section above). 
<details>
<summary>Should we make the search input and button text font family be the same?</summary>
<img width="1024" alt="Screenshot 2022-06-30 at 23 21 58" src="https://user-images.githubusercontent.com/25318659/176771779-09720952-419a-4993-88c7-9c2116cfedc4.png">

</details>

> Question: A question for the engineers who use IntelliJ. What do we think about not using the Sourcegraph style "Tab button and just having the helper text on the keyboard shortcut directly to the right in bold?

<details>
<summary>Good idea! Could it be something like this 👇🏻 ?</summary>

<table>
<tr>
<td>
<img width="1024" alt="Screenshot 2022-06-30 at 23 20 54" src="https://user-images.githubusercontent.com/25318659/176771464-f49c8e48-b09a-4cf4-b504-4ac49ba59014.png">
</td>
<td>
<img width="1024" alt="Screenshot 2022-06-30 at 23 03 53" src="https://user-images.githubusercontent.com/25318659/176768075-954057ab-35af-4464-a018-315db89639ef.png">
</td>
</tr>
</table>
</details>

I'm still not sure about the search context dropdown and input styles. Do we have a Figma describing the details or do we style them compared to the JetBrains search popup?

[Loom](https://www.loom.com/share/9bcee43fe3d94566a70b11c45145102e)

vdavid commented:
@taras-yemets 

> I somehow missed this option, thank you! Implemented the match query highlight in yellow. 
>
<details>
<summary><b>Screenshots</b></summary>
<table>
<tr>
<td>
<img width="1024" alt="Screenshot 2022-06-30 at 23 14 04" src="https://user-images.githubusercontent.com/25318659/176770322-a15231c6-6c97-426f-9c18-3c089fb01bad.png">
</td>
<td>
<img width="1024" alt="Screenshot 2022-06-30 at 23 14 27" src="https://user-images.githubusercontent.com/25318659/176770373-c46478b2-d497-4f83-958c-5a91f905763e.png">
</td>
</tr>
</table>
</details>

I like the direction!
In the screenshots, the highlights looked different to me from the search result list, so I checked how they looked in JetBrains' version.
Turns out it looks like this, different from both the highlights in the screenshot and from the highlights we use in the search results:
<img width="779" alt="CleanShot 2022-06-30 at 22 56 22@2x" src="https://user-images.githubusercontent.com/2552265/176776941-b829dff6-9a8d-47a7-b5ed-ebddcc455a51.png">
They actually use a vertical gradient for the yellow background. 🤷‍♂️ Any chance we can nail matching this pretty much perfectly? 😬

taras-yemets commented:
Thank you for your comment, @vdavid! 

> They actually use a vertical gradient for the yellow background. :man-shrugging: Any chance we can nail matching this pretty much perfectly?

I wonder if there is a way to inspect JetBrains native UI elements (e.g. dropdowns) styles. I was looking for colors on _localhost:3000/?color-debug_ and couldn't find the exact same color for the text highlight, so I used [similar ones](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@taras-yemets/jetbrains-adjust-dropdown-styles/-/blob/client/jetbrains/webview/src/search/index.tsx?L112) (search results component uses [hardcoded colors](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@taras-yemets/jetbrains-adjust-dropdown-styles/-/blob/client/jetbrains/webview/src/search/results/FileSearchResult.module.scss?L8-9)).
How did you fix the styling issues previously? Is there a Figma with plugin styles or do you visually compare plugin with the native UI elements?

cc: @philipp-spiess

philipp-spiess commented:
Here's a screenshot of the light mode so it's clearer what I meant with the differences:

![Screenshot 2022-07-01 at 10 36 07](https://user-images.githubusercontent.com/458591/176857975-dc4dd2c7-1bde-4057-95d7-2011b3a20042.png)

This is what I based my hardcoded values on

taras-yemets commented:
Thank you, @philipp-spiess! Makes sense to me.

For dropdown matches I'll use the same hardcoded colors you used for the search results. Then we can have another iteration on colors adjustment (in a separate PR).

taras-yemets commented:
> For dropdown matches I'll use the same hardcoded colors you used for the search results.

Updated in https://github.com/sourcegraph/sourcegraph/pull/37833/commits/987b5378a419faea036bb4bc4dd8228344f15df1

<details>
<summary><b>Screenshots</b></summary>
<table>
<tr>
<td>
<img width="1024" alt="Screenshot 2022-07-01 at 15 10 58" src="https://user-images.githubusercontent.com/25318659/176892191-0528b630-f9e1-49be-a7fc-b61b6299a173.png">
</td>
<td>
<img width="1024" alt="Screenshot 2022-07-01 at 15 11 24" src="https://user-images.githubusercontent.com/25318659/176892206-65df7041-a929-4c87-8765-4e5d2a84c518.png">

</td>
</tr>
</table>
</details>

taras-yemets commented:
Thanks for your review, @jjinnii!

<details>
<summary>Changes made since your last review</summary>
<table>
<thead>
<tr>
<td>Element</td>
<td>Description</td>
<td>Screenshots</td>
<td>JetBrains example</td>
</tr>
</thead>
<tbody>
<tr>
<td>Search input and auto-complete menu</td>
<td>
  <ul>
    <li>use default font-family for input and list items</li>
    <li>reduce item height to 20px</li>
    <li>align completion detail text to the right (inspired by JetBrains global search menu)</li>
    <li>move autocompletion hint ("Tab") to the menu footer (inspired by JetBrains "Search everywhere" menu)</li>
    <li>remove menu border radius</li>
    <li>adjust border color (theme variable, although I'm not sure whether it's the correct one) and box shadow (hardcoded)</li>
    </ul>
</td>
<td>
  
<img width="995" alt="Screenshot 2022-07-05 at 13 41 01" src="https://user-images.githubusercontent.com/25318659/177310436-47eaf32f-102a-4131-a2b1-bd60362de7b0.png">
<img width="995" alt="Screenshot 2022-07-05 at 13 40 46" src="https://user-images.githubusercontent.com/25318659/177310446-ff5b0b37-7a80-4522-8d82-7fc0a46bd436.png">

</td>
<td>
<img width="692" alt="Screenshot 2022-07-05 at 14 08 48" src="https://user-images.githubusercontent.com/25318659/177314641-e3bee1cb-3fe6-48d0-948e-cc0a6cc5669c.png">

<img width="705" alt="Screenshot 2022-07-05 at 14 08 24" src="https://user-images.githubusercontent.com/25318659/177314588-794e8807-929d-47ad-917d-d850f0d58c05.png">

</td>
</tr>
<tr>
<td>Search context button and menu</td>
<td>
<ul>
<li>use default font-family for button, input and menu items</li>
<li>highlight menu header and footer (inspired by JetBrains global search menu)</li>
<li>remove horizontal padding on input container so that input takes the whole menu width (inspired by global search, "Search anywhere" menus inputs)</li>
<li>reduce menu items height to 20px</li>
<li>adjust menu items colors</li>
<li>change reset button styles and move it to the right (inspired by CTA buttons in JetBrains global search, "Search anywhere" menus)</li>
<li>adjusted border color and box shadow </li>
<li>border color and box-shadow: the same as for the search auto-complete menu</li>
<li>adjust focused button border and box shadow</li>
</ul>
</td>
<td>
<img width="995" alt="Screenshot 2022-07-05 at 13 59 13" src="https://user-images.githubusercontent.com/25318659/177313292-c7e33ce5-1072-4cec-831c-2567c9a8c819.png">

<img width="995" alt="Screenshot 2022-07-05 at 13 59 27" src="https://user-images.githubusercontent.com/25318659/177313277-16fc9d8f-da1c-4bb8-a486-cc62864dd1bb.png">
<img width="285" alt="Screenshot 2022-07-05 at 14 02 20" src="https://user-images.githubusercontent.com/25318659/177313700-a8785ff0-fc4a-4f97-8c75-9cfcd39a5265.png">
<img width="285" alt="Screenshot 2022-07-05 at 14 02 09" src="https://user-images.githubusercontent.com/25318659/177313705-aa81d87e-24c5-4eed-880e-786d02d93d5c.png">
<img width="285" alt="Screenshot 2022-07-05 at 14 01 55" src="https://user-images.githubusercontent.com/25318659/177313707-c3734053-5465-4bd4-92d6-b8f5338eecc4.png">
<img width="285" alt="Screenshot 2022-07-05 at 14 01 28" src="https://user-images.githubusercontent.com/25318659/177313708-84ca48a4-ac55-4c60-bfee-5a754f94af7d.png">



</td>
<td>
<img width="924" alt="Screenshot 2022-07-05 at 14 10 09" src="https://user-images.githubusercontent.com/25318659/177314910-0c64fd54-b600-4c51-983c-171f6ec833d4.png">
<img width="930" alt="Screenshot 2022-07-05 at 14 10 33" src="https://user-images.githubusercontent.com/25318659/177314922-fc5bda0c-f7bb-4f21-b54c-ac7889d2209b.png">
<img width="243" alt="Screenshot 2022-07-05 at 14 11 43" src="https://user-images.githubusercontent.com/25318659/177315036-9fef1956-81a7-4bce-ae70-09923214a619.png">
<img width="209" alt="Screenshot 2022-07-05 at 14 11 32" src="https://user-images.githubusercontent.com/25318659/177315046-85e4f300-c91a-41bf-bd13-3569f5a3e306.png">


</td>
</tr>
</tbody>
</table>
</details>

What do you think about the changes above? Do the dropdown items look less condensed than the native JetBrains dropdown items? Which UI elements would you like to improve?

jjinnii commented:
@taras-yemets This looks amazing!! 🎉 🎉  The auto-complete menu in particular looks _miles_ better. I have a few small suggestions for the context menu, but I think it's less important than the input field, so I'll put those as low priority in another issue. 😄 

I think the only improvement that I'd like to see from this issue is to update the button for the contexts button as well as the "reset" button in the context menu to match the styling of buttons in Jetbrains. (the height of the button + rounded corners) We've styled this button in #38081 for reference. 

<img width="995" alt="177313292-c7e33ce5-1072-4cec-831c-2567c9a8c819" src="https://user-images.githubusercontent.com/87138876/177318896-d6f42e3b-5680-4844-8154-b33bbf281f7c.png">

<img width="866" alt="Screen Shot 2022-07-05 at 8 37 24 PM" src="https://user-images.githubusercontent.com/87138876/177318980-e1e81db3-143f-42bc-bc4b-e3480125b134.png">

jjinnii reviewed (approved):
Preemptively approving, as I only have that one small suggestion.

jjinnii commented:
FYI @taras-yemets  you can inspect the styling of _some_ JetBrains components in their UI kit Figma file. Here is the section for[ buttons](https://www.figma.com/file/uNnoEQqEJepQKUUbdXvTuQ/IntelliJ-Platform-UI-Kit-(Community)?node-id=5%3A0), which has a height of 24 px or 1.5 rem.
